soc/ti/simplelink/cc23x0/power.c

```diff
@@ -0,0 +1,245 @@
+
+	if (rtc_mask != 0) {
+		if (rtc_mask & 0x1) {
+			HWREG(RTC_BASE + RTC_O_CH0CC8U) = rtc[0];
+		}
+		if (rtc_mask & 0x2) {
+			HWREG(RTC_BASE + RTC_O_CH1CC8U) = rtc[1];
+		}
+	}
+
+	LRFDApplyClockDependencies();
+	PowerCC23X0_notify(PowerLPF3_AWAKE_STANDBY);
+	HwiP_restore(key);
+}
+#endif
+
+static void pm_cc23x0_enter_standby(void)
+{
+#ifndef CONFIG_CC23X0_RTC_TIMER
+	uint32_t rtc_now = 0;
+	uint32_t systim_now = 0;
+	uint32_t systim_next = MAX_SYSTIMER_DELTA;
+	uint32_t systim_delta = 0;
+	uint32_t rtc_delta = 0;
+
+	key = HwiP_disable();
+
+	uint32_t constraints = Power_getConstraintMask();
+	bool standby = (constraints & (1 << PowerLPF3_DISALLOW_STANDBY)) == 0;
+	bool idle = (constraints & (1 << PowerLPF3_DISALLOW_IDLE)) == 0;
+
+	if (standby && (HWREG(CKMD_BASE + CKMD_O_LFCLKSEL) & CKMD_LFCLKSEL_MAIN_LFOSC) &&
+	    !(HWREG(CKMD_BASE + CKMD_O_LFCLKSTAT) & CKMD_LFCLKSTAT_FLTSETTLED_M)) {
+		standby = false;
+		idle = false;
+	}
+
+	if (standby) {
+		systim_mask = HWREG(SYSTIM_BASE + SYSTIM_O_IMASK);
+		if (systim_mask != 0) {
+			systim_next = 0xFFFFFFFF;
+			systim_now = HWREG(SYSTIM_BASE + SYSTIM_O_TIME1U);
+			for (uint8_t idx = 0; idx < SYSTIM_CH_CNT; idx++) {
+				if (systim_mask & (1 << idx)) {
+					systim[idx] = HWREG(SYSTIM_BASE + SYSTIM_CH(idx));
+					systim_delta = systim[idx];
+					systim_delta -= systim_now << systim_offset[idx];
+
+					if (systim_delta > MAX_SYSTIMER_DELTA) {
+						systim_delta = 0;
+					}
+
+					systim_delta = systim_delta >> systim_offset[idx];
+					systim_next = MIN(systim_next, systim_delta);
+				}
+			}
+		} else {
+			systim_next = MAX_SYSTIMER_DELTA;
+		}
+
+		rtc_mask = HWREG(RTC_BASE + RTC_O_IMASK);
+		if (rtc_mask != 0) {
+			rtc_now = HWREG(RTC_BASE + RTC_O_TIME8U);
+			if (rtc_mask & 0x1) {
+				rtc[0] = HWREG(RTC_BASE + RTC_O_CH0CC8U);
+				rtc_delta = (rtc[0] - rtc_now) << 3;
+				systim_next = MIN(systim_next, rtc_delta);
+			}
+			if (rtc_mask & 0x2) {
+				rtc[1] = HWREG(RTC_BASE + RTC_O_CH1CC8U);
+				rtc_delta = (rtc[1] - rtc_now) << 3;
+				systim_next = MIN(systim_next, rtc_delta);
+			}
+		}
+
+		if (systim_next > PowerCC23X0_TOTALTIMESTANDBY) {
+			HWREG(EVTSVT_BASE + EVTSVT_O_CPUIRQ3SEL) = 0;
+			HWREG(EVTSVT_BASE + EVTSVT_O_CPUIRQ16SEL) =
+			      EVTSVT_CPUIRQ16SEL_PUBID_AON_RTC_COMB;
+			HwiP_clearInterrupt(INT_CPUIRQ16);
+			rtc_now = HWREG(RTC_BASE + RTC_O_TIME8U);
+			HWREG(RTC_BASE + RTC_O_CH0CC8U) = RTC_NEXT(systim_next, rtc_now);
+#endif
+			Power_sleep(PowerLPF3_STANDBY);
+#ifndef CONFIG_CC23X0_RTC_TIMER
+			pm_cc23x0_systim_standby_restore();
+		} else if (idle) {
+			__WFI();
+		}
+	} else if (idle) {
+		__WFI();
+	}
+
+	HwiP_restore(key);
+#endif
+}
+
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+
+	switch (state) {
+	case PM_STATE_RUNTIME_IDLE:
+		PowerCC23X0_doWFI();
+		break;
+	case PM_STATE_STANDBY:
+		pm_cc23x0_enter_standby();
+		break;
+	case PM_STATE_SOFT_OFF:
+		Power_shutdown(0, 0);
+		break;
```

Comment:
Since this is shutdown, wouldn't it make more sense to implement z_sys_poweroff instead of having it as a power mode?

Because, after going into this state, as far as I understand the datasheet, you wake up by going through a whole boot cycle. So shutdown seems more appropriate from my side.

Reply:
I have the assumption, that the function `Power_shutdown(0, 0);` will shut down the CPU. Meaning, it will not resume from the shutdown in this function with the RAM content intact, but instead do a full boot cycle afterwards if an interrupt wakes it up.

If my assumption holds up, then this is not part of power-management. Power management should the the automatic power management, that saves power by going to sleep and resume execution as if nothing happened in between.

There is another API for powering down the CPU, which is to implement the `z_sys_poweroff` function for the CPU. If you do "git grep z_sys_poweroff" you will see many [examples](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/ti/mspm0/common/poweroff.c) for it.

Reply:
No, you are right, I was not aware of that. It is now just that I don't like that behaviour and there is not much in the codebase, that does it that way. Most go for the z_sys_poweroff implementation, which I find cleaner.